client/go/dogma/watch_service.go

```diff
@@ -0,0 +1,378 @@
+		var value interface{}
+		if convertingValueFunc != nil {
+			value = convertingValueFunc(result.Commit.Entries[0].Content)
+		} else {
+			value = result.Commit.Entries[0].Content
+		}
+
+		return &Latest{Revision: result.Commit.Revision, Value: value}
+	}
+	return w, nil
+}
+
+func (ws *watchService) repoWatcher(projectName, repoName, pathPattern string,
+	convertingValueFunc func(revision int) interface{}) (*Watcher, error) {
+	w := newWatcher(projectName, repoName, pathPattern)
+	w.doWatchFunc = func(lastKnownRevision int) <-chan *WatchResult {
+		return ws.watchRepo(w.watchCTX, projectName, repoName, strconv.Itoa(lastKnownRevision),
+			pathPattern, watchTimeout)
+	}
+	w.convertingResultFunc = func(result *WatchResult) *Latest {
+		var value interface{}
+		revision := result.Commit.Revision
+		if convertingValueFunc != nil {
+			value = convertingValueFunc(revision)
+		} else {
+			value = revision
+		}
+
+		return &Latest{Revision: revision, Value: value}
+	}
+	return w, nil
+}
+
+func (w *Watcher) start() {
+	if atomic.CompareAndSwapInt32(&w.state, initial, started) {
+		w.scheduleWatch(0)
+	}
+}
+
+func (w *Watcher) scheduleWatch(numAttemptsSoFar int) {
+	if w.isStopped() {
+		return
+	}
+
+	var delay time.Duration
+	if numAttemptsSoFar == 0 {
+		latest := w.Latest()
+		if latest.Err != nil {
+			delay = delayOnSuccess
+		} else {
+			delay = 0
+		}
+	} else {
+		delay = nextDelay(numAttemptsSoFar)
+	}
+
+	go func() {
+		select {
+		case <-w.scheduleCTX.Done():
+		case <-time.NewTimer(delay).C:
+			w.doWatch(numAttemptsSoFar)
+		}
+	}()
+}
+
+func (w *Watcher) isStopped() bool {
+	state := atomic.LoadInt32(&w.state)
+	return state == stopped
+}
+
+func (w *Watcher) doWatch(numAttemptsSoFar int) {
+	if w.isStopped() {
+		return
+	}
+
+	var lastKnownRevision int
+	curLatest := w.Latest()
+	if curLatest.Revision == 0 {
+		lastKnownRevision = 1 // Init revision
+	} else {
+		lastKnownRevision = curLatest.Revision
+	}
+
+	select {
+	case <-w.watchCTX.Done():
+	case watchResult := <-w.doWatchFunc(lastKnownRevision):
+		if watchResult.Err != nil {
+			if watchResult.Err == context.Canceled {
+				// Cancelled by close()
+				return
+			}
+
+			log.Debug(watchResult.Err)
+			w.scheduleWatch(numAttemptsSoFar + 1)
+			return
+		}
+
+		newLatest := w.convertingResultFunc(watchResult)
+		if oldLatest := w.Latest(); oldLatest.Err != nil {
+			// latest is not set yet
+			w.initialChan <- newLatest
+		}
+		atomic.StorePointer(&w.latestPtr, (unsafe.Pointer)(newLatest))
+		log.Debugf("Watcher noticed updated file %s/%s%s: rev=%v",
+			w.projectName, w.repoName, w.pathPattern, newLatest.Revision)
+		w.notifyListeners()
+		w.scheduleWatch(0)
+	}
+}
+
+func (w *Watcher) notifyListeners() {
+	if w.isStopped() {
+		// Do not notify after stopped.
+		return
+	}
+
+	latest := w.Latest()
+	w.listenersMutex.Lock()
+	defer w.listenersMutex.Unlock()
+
+	go func() {
+		for _, listener := range w.updateListeners {
```

Comment:
Doesn't this goroutine get executed later after `listernsMutext.Unlock()` is called and `notifyListeners()` return? Shouldn't we make a copy of the listeners to make this really thread-safe?

Reply:
Oops I will make a copy of it. Originally there was no Goroutine that executes the `listeners`.
But I thought that using Goroutine here is better in case something happens while executing `listerens` such as long time job.
So I put it and forgot to make a copy of it. :-)